src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/Huffman.cs

```diff
@@ -269,36 +273,127 @@ internal class Huffman
-            (24, new[] { 9, 142, 144, 145, 148, 159, 171, 206, 215, 225, 236, 237 }),
-            (25, new[] { 199, 207, 234, 235 }),
-            (26, new[] { 192, 193, 200, 201, 202, 205, 210, 213, 218, 219, 238, 240, 242, 243, 255 }),
-            (27, new[] { 203, 204, 211, 212, 214, 221, 222, 223, 241, 244, 245, 246, 247, 248, 250, 251, 252, 253, 254 }),
-            (28, new[] { 2, 3, 4, 5, 6, 7, 8, 11, 12, 14, 15, 16, 17, 18, 19, 20, 21, 23, 24, 25, 26, 27, 28, 29, 30, 31, 127, 220, 249 }),
-            (30, new[] { 10, 13, 22, 256 })
-        };
-
         public static (uint encoded, int bitLength) Encode(int data)
         {
             return _encodingTable[data];
         }
 
+        private static ushort[] GenerateDecodingLookupTree()
+        {
+            // Decoding lookup tree is a tree of 8 bit lookup tables stored in
+            // one dimensional array of ushort to reduce allocations.
+            // First 256 ushort is lookup table with index 0, next 256 ushort is lookup table with index 1, etc...
+            // loopk_value = [(lookup_table_index << 8) + lookup_index]
+
+            // lookup_index is next 8 bits of huffman code, if there is less than 8 bits in source.
+            // lookup_index MUST be aligned to 8 bits with LSB bits set to anything (zeros are recommended).
+
+            // Lookup value is encoded in ushort as either.
+            // -----------------------------------------------------------------
+            //  15  14  13  12  11  10   9   8   7   6   5   4   3   2   1   0
+            // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+            // | 1 |   next_lookup_table_index |            not_used           |
+            // +---+---------------------------+-------------------------------+
+            // or
+            // +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+            // | 0 |     number_of_used_bits   |              octet            |
+            // +---+---------------------------+-------------------------------+
+
+            // Bit 15 unset indicates a leaf value of decoding tree.
+            // For example value 0x0241 means that we have reached end of huffman code
+            // with result byte 0x41 'A' and from lookup bits only rightmost 2 bits were used
+            // and rest of bits are part of next huffman code.
+
+            // Bit 15 set indicates that code is not yet decoded and next lookup table index shall be used
+            // for next n bits of huffman code.
+            // 0 in 'next lookup table index' is considered as decoding error - invalid huffman code
+
+            // Because HPack uses static huffman code defined in RFC https://httpwg.org/specs/rfc7541.html#huffman.code
+            // it is guaranteed that for this huffman code generated decoding lookup tree MUST consist of exactly 15 lookup tables
+            var decodingTree = new ushort[15 * 256];
```

Comment:
Actually you can just use 2-dimensional, rectangular array: `new ushort[15, 256]`. And then indexed access of `decodingTree[lookupTableIndex, (indexInLookupTable | suffix)]`. It will be represented in memory as single dim array and compiler/runtime will handle the offset calculations for you.

Reply:
Tempting, I will verify performance of this and if it will be same, and I expect it will, I will change it. 
To be continued...

Reply:
> 2-dimensional, rectangular array

The runtime can't handle these as good as sz-arrays (see e.g. https://github.com/dotnet/runtime/issues/35293, https://github.com/dotnet/runtime/issues/5481) and indexing is mostly method-based as opposed to direct (JIT) support for the sz-arrays.

So I doubt the perf with the rectangular array can be on par.

Reply:
Benchmark:
|      Method | DataIndex |         Mean |     Error |    StdDev |         Op/s | Ratio |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------ |---------- |-------------:|----------:|----------:|-------------:|------:|-------:|------:|------:|----------:|
|     Current |        -1 |    575.93 ns |  4.116 ns |  3.648 ns |  1,736,315.4 |  1.00 |      - |     - |     - |      64 B |
|      Tree2d |        -1 |    598.66 ns |  6.873 ns |  6.093 ns |  1,670,391.6 |  1.04 |      - |     - |     - |      64 B |

It is slightly slower. I believe it is because my custom index computing using fast 'x<<8' is faster than x*n when n is not known at compile time thus had to be read from array header.

On the other hand, it is slightly more readable with ushort[,]. But since array index computation is used only at about 5 places, I'd recommend to keep it as is.

What do you think?

Reply:
I'd leave it as is -- same as you said.

Another (minor) point is that the current implementation (sz-array) produces less JITed code, which is preferable (in most cases).
For readability current version is readable -- I even see no need to extract the index-computation to a (local) function as this is just overhead (for the JIT to inline this function).

Reply:
Hmm. As for index computation I though compiler would inline it, wouldn't it?
I can inline it, it would not harm readability much. @gfoidl Shall I?

Reply:
> for index computation I though compiler would inline it,

Yeah, it will inline it. It's just a shift and add, so pretty small and straight-forward. As it's so "mini" I'd leave it manually inlined. Also the logic isn't very complicated.

I would extract it to a separate method if unsafe indexing is used, so that Debug.Asserts can be used to verify the access is within bounds. But this is not the case here. Just leave it as is.